Step1 자바 reflection

sehan opened the pull request:
자바 reflection 리뷰 요청드립니다. 

잘 부탁드립니다. 🥰

nio-d committed:
feat: 클래스 정보출력

nio-d committed:
feat: test 로 시작하는 메소드 실행

nio-d committed:
feat: @Test 애노테이션 메소드 실행

nio-d committed:
refactor: isAnnotationPresent 사용해서 @MyTest 존재여부체크

nio-d committed:
feat: private field에 값 할당

nio-d committed:
feat: 인자를 가진 생성자의 인스턴스 생성

nio-d committed:
feat: component scan

nio-d committed:
chore: 불필요한 코드삭제

karian7 commented on src/test/java/next/reflection/ReflectionTest.java:
```diff
@@ -5,14 +5,65 @@
 public class ReflectionTest {
     private static final Logger logger = LoggerFactory.getLogger(ReflectionTest.class);
 
+    enum Modifier {
```
enum 을 사용하신것에 대해 요구사항 달성을 위해 빠르고 간편한 방법이긴 하지만 api 활용연습인 만큼 [java.lang.reflect.Modifier](https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/Modifier.html ) 클래스를 활용해 보시는건 어떨까요?

sehan replied:
java.lang.reflect.Modifier 가 있었군요. 이걸 사용하는게 좋겠네요 👍

karian7 commented on src/test/java/next/reflection/ReflectionTest.java:
```diff
@@ -5,14 +5,65 @@
+                    logger.debug(MessageFormat.format("constructors: {0} {1} ({2})",
+                            Modifier.valueOf(constructor.getModifiers()).name(),
+                            clazz.getName(),
+                            Arrays.stream(constructor.getParameterTypes()).map(it -> it.getName()).collect(Collectors.joining(" ,"))));
```
```suggestion
                            Arrays.stream(constructor.getParameterTypes()).map(it -> it.getName()).collect(Collectors.joining(", "))));
```
[의견] ~조금 짜치지만~ 이게 맞을것 같아요 😅

sehan replied:
네, 그게 맞네요 😅

karian7 commented on src/test/java/next/reflection/ReflectionTest.java:
```diff
@@ -5,14 +5,65 @@
+        // 생성자
+        Arrays.stream(clazz.getDeclaredConstructors())
+                .forEach(constructor -> {
+                    logger.debug(MessageFormat.format("constructors: {0} {1} ({2})",
```
`MessageFormat` 👍

sehan replied:
> 안녕하세요? 니오. 리뷰를 맡게된 킷입니다. 이곳에서 이렇게 만나게 되네요 😃😃
> 리뷰가 좀 늦었네요. 죄송해요 🙏
> 
> 첫번째 과제 잘 수행해 주셨습니다. 💯💯
> 몇가지 코맨트 남겼는데요 `[의견]` 이 붙은 코맨트는 필수로 코드를 수정할 필요가 없는 부분이니 참고 해 주세요.
> 
> 도전과제를 하나 더 드려볼께요. 이부분 역시 바쁘시면 패스하셔도 됩니다. 😎
> 
> * `BeanFactoryTest#di` 테스트가 성공하도록 `BeanFactory` 개선해보기
> 
> 그럼 잘 부탁드려요~ 💪 🚀

test 를 돌릴때 마다 BeanFactoryTest#di 깨지는걸 보고 이걸 fix 해야 하나 고민하다가 고이고이 접어 넣어두었는데 이렇게 다시 등장하네요. 😄  가능한 노력해보겠습니다.

karian7 commented on src/test/java/next/reflection/Junit3TestRunner.java:
```diff
@@ -2,10 +2,34 @@
+            } catch (NoSuchMethodException e) {
+                e.printStackTrace();
+            }
+        });
```
[의견] 이경우엔 for-each 스타일이 좀 더 간결할 것 같아요.

sehan replied:
네, 저도 이건 코드를 볼때 마다 마음이 왔다갔다 하네요. 

forEach 를 사용하면 변수에서 .forEach 매소드로 코드전개(타이핑)이 이어지는 느낌이 좋아서 최근에는 이렇게 작성하고 있어요.

for( T item : items ) -> "for 를 돌릴건데 items 에서 하나씩 꺼내서 item 에 담아줄게 하나씩 처리해줘"
items.forEach( action ) -> "items 을 for 처리할 거야 action 이 하나씩 처리할거야"

for( T item : items ) 는 흐름이 <---- 이렇게 흘러가는 느낌이고 items.forEach( action ) 은 -----> 이렇게 흘러가는 느낌이랄까요?
흔들리는 마음은 갈피를 못잡고 있네요 😄

karian7 replied:
아, 이유를 말씀드리지 않았네요.
제가 for-each 가 더 간결하다고 생각한 부분은 이 메소드 시그니처가 `throws Exception` 이기 때문이었어요. 또한 테스트의 경우 예외를 catch 하는것 보다 throw 시키는게 더 좋지 않을까? 하는 생각에서 나온 의견이었습니다.
https://github.com/next-step/jwp-mvc-kakao/pull/2/commits/f768b52e336db70cbe282659de6d1c072c4af2e5 에서 적용해주셨군요! 반영 감사합니다~ 👍

karian7 commented on src/test/java/next/reflection/Junit4TestRunner.java:
```diff
@@ -15,8 +15,7 @@ public void run() throws Exception {
         List<Method> annotatedMethods = Arrays.stream(clazz.getDeclaredMethods())
-                .filter(method -> Arrays.stream(method.getAnnotations())
-                        .anyMatch(annotation -> annotation.annotationType().getSimpleName().equals("MyTest")))
+                .filter(method -> method.isAnnotationPresent(MyTest.class))
```
`isAnnotationPresent` 👍

karian7 commented on src/test/java/next/reflection/ReflectionTest.java:
```diff
@@ -102,4 +105,19 @@ public void privateFieldAccess() throws NoSuchMethodException, IllegalAccessExce
+                .isEqualTo(new Question("nio", "title", "contents"));
+
+        Constructor constructor2 = clazz.getDeclaredConstructor(long.class, String.class, String.class, String.class, Date.class, int.class);
+        assertThat(constructor2.newInstance(1L, "nio", "title", "contents", new Date(), 0))
```
```suggestion
        assertThat(constructor2.newInstance(1L, "nio", "title", "contents", new Date(), 1))
```
이렇게 변경해도 테스트가 통과합니다. 원인을 파악해주시고 `Question` 클래스의 수정없이 검증을 해보시겠어요?

sehan replied:
테스트가 통과하는건 questionId 가 동일하면 equals 로 판단하고 있어서 다른 프로퍼티들과 관계 없이 통과합니다. 😅 

생성자가 정상적으로 호출되었는지를 테스트하기에 인스턴스 비교는 부적절 하겠네요. 프로퍼티를 체크하는 형태로 수정하도록 하겠습니다.

karian7 reviewed (changes requested):
안녕하세요? 니오. 리뷰를 맡게된 킷입니다. 이곳에서 이렇게 만나게 되네요 😃😃 
리뷰가 좀 늦었네요. 죄송해요 🙏 

첫번째 과제 잘 수행해 주셨습니다. 💯💯 
몇가지 코맨트 남겼는데요 `[의견]` 이 붙은 코맨트는 필수로 코드를 수정할 필요가 없는 부분이니 참고 해 주세요.

도전과제를 하나 더 드려볼께요. 이부분 역시 바쁘시면 패스하셔도 됩니다. 😎 
* `BeanFactoryTest#di` 테스트가 성공하도록 `BeanFactory` 개선해보기

그럼 잘 부탁드려요~ 💪 🚀

nio-d committed:
refactor: pr 피드백 반영

> java.lang.reflect.Modifier 사용하기
> 생성자 시그니쳐 출력포멧 다듬기
> constructor 호출 검증테스트 개선

nio-d committed:
test: BeanFactory#di test 추가

karian7 commented on src/main/java/core/di/factory/BeanFactory.java:
```diff
@@ -23,6 +24,43 @@ public BeanFactory(Set<Class<?>> preInstanticateBeans) {
+            if (injectedConstructor != null) {
+                List<Object> params = new ArrayList<>();
+                for( Class paramType : injectedConstructor.getParameterTypes()){
+                    params.add( createInstance(paramType) );
```
생성자 주입을 어떻게 풀어야 하나 고민되었는데 재귀로 풀어내셨네요 👍 💯

karian7 replied:
생성자 주입을 어떻게 풀어야 하나 고민되었는데 재귀로 풀어내셨네요 👍 💯

karian7 commented on src/test/java/next/reflection/ReflectionTest.java:
```diff
@@ -109,13 +91,26 @@ public void privateFieldAccess() throws NoSuchMethodException, IllegalAccessExce
+                .hasFieldOrPropertyWithValue("title", "title")
+                .hasFieldOrPropertyWithValue("contents", "contents")
+                .hasFieldOrPropertyWithValue("createdDate", now)
+                .hasFieldOrPropertyWithValue("countOfComment", 0);
```
피드백 반영해주셔서 감사합니다. 저는 이렇게 해보는것도 좋겠다고 생각했어요. ☺️
```java
assertThat(q2)
                .usingRecursiveComparison()
                .isEqualTo(new Question(1L, "nio", "title", "contents", now, 0));
```

sehan replied:
usingRecursiveComparison() 이런 기능도 있었군요.  유용하게 활용할수 있겠네요.  감사합니다. 😄

karian7 reviewed (approved):
모든 피드백을 반영해 주셨군요! 🥰
di 쪽은 저도 어떻게 될지 궁금했는데 `BeanFactoryUtils` 같은 기존에 제공되는 코드를 활용해서 잘 구현해 주신 것 같아요 💯💯

머지하겠습니다. 다음 단계들도 화이팅 입니다!
